pkg/ccl/changefeedccl/schemafeed/schema_feed.go

```diff
@@ -481,27 +470,45 @@ func (tf *schemaFeed) pauseOrResumePolling(
 		return atOrBefore, nil
 	}
 
-	if tf.mu.allTableVersions1 == nil {
-		tf.mu.allTableVersions1 = make(map[descpb.ID]descpb.DescriptorVersion)
-		tf.mu.allTableVersions2 = make(map[descpb.ID]descpb.DescriptorVersion)
-	}
-
 	// Always start with a stance to resume polling until we've proved otherwise.
 	tf.mu.pollingPaused = false
-	if ok, err := areAllLeasedTablesSchemaLockedAt(tf.mu.highWater, tf.mu.allTableVersions1); err != nil || !ok {
-		return atOrBefore, err
-	}
-	if ok, err := areAllLeasedTablesSchemaLockedAt(atOrBefore, tf.mu.allTableVersions2); err != nil || !ok {
-		return atOrBefore, err
-	}
-	if len(tf.mu.allTableVersions1) != len(tf.mu.allTableVersions2) {
-		return atOrBefore, nil
-	}
-	for id, version := range tf.mu.allTableVersions1 {
-		if version != tf.mu.allTableVersions2[id] {
-			return atOrBefore, nil
+
+	// Determine if table descriptor versions at highWater and at atOrBefore are
+	// both locked, and have the same versions.
+	allSame, err := func() (bool, error) {
+		for _, id := range tf.tableIDs {
+			var highWaterVersion descpb.DescriptorVersion
+			if tf.mu.versionsLockedAt.Less(tf.mu.highWater) {
+				// Need to re-load locked versions at highWater.
+				v, isLocked, err := tableDescriptorStateAt(id, tf.mu.highWater)
+				if err != nil || !isLocked {
+					return false, err
+				}
+				tf.mu.lockedVersions[id] = v
+				highWaterVersion = v
+				tf.mu.versionsLockedAt = tf.mu.highWater
+			} else {
```

Comment:
nit: drop the else block and also drop `highWaterVersion = v`. We can do this after the if:
```
highWaterVersion = tf.mu.lockedVersions[id]
```